engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAnyNode.java

```diff
@@ -0,0 +1,609 @@
+      @CachedLibrary("selfDateTime") InteropLibrary selfInterop,
+      @CachedLibrary("otherDateTime") InteropLibrary otherInterop) {
+    LocalDateTime selfJavaDateTime;
+    LocalDateTime otherJavaDateTime;
+    try {
+      selfJavaDateTime = LocalDateTime.of(
+          selfInterop.asDate(selfDateTime),
+          selfInterop.asTime(selfDateTime)
+      );
+      otherJavaDateTime = LocalDateTime.of(
+          otherInterop.asDate(otherDateTime),
+          otherInterop.asTime(otherDateTime)
+      );
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+    return selfJavaDateTime.isEqual(otherJavaDateTime);
+  }
+
+  @Specialization(guards = {
+      "selfInterop.isDate(selfDate)",
+      "!selfInterop.isTime(selfDate)",
+      "otherInterop.isDate(otherDate)",
+      "!otherInterop.isTime(otherDate)"
+  }, limit = "3")
+  boolean equalsDates(Object selfDate, Object otherDate,
+      @CachedLibrary("selfDate") InteropLibrary selfInterop,
+      @CachedLibrary("otherDate") InteropLibrary otherInterop) {
+    LocalDate otherJavaDate;
+    LocalDate selfJavaDate;
+    try {
+      otherJavaDate = otherInterop.asDate(otherDate);
+      selfJavaDate = selfInterop.asDate(selfDate);
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+    return selfJavaDate.isEqual(otherJavaDate);
+  }
+
+  @Specialization(guards = {
+      "!selfInterop.isDate(selfTime)",
+      "selfInterop.isTime(selfTime)",
+      "!otherInterop.isDate(otherTime)",
+      "otherInterop.isTime(otherTime)"
+  }, limit = "3")
+  boolean equalsTimes(Object selfTime, Object otherTime,
+      @CachedLibrary("selfTime") InteropLibrary selfInterop,
+      @CachedLibrary("otherTime") InteropLibrary otherInterop) {
+    LocalTime selfJavaTime;
+    LocalTime otherJavaTime;
+    try {
+      selfJavaTime = selfInterop.asTime(selfTime);
+      otherJavaTime = otherInterop.asTime(otherTime);
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+    return selfJavaTime.equals(otherJavaTime);
+  }
+
+  @Specialization(guards = {
+      "selfInterop.isDuration(selfDuration)",
+      "otherInterop.isDuration(otherDuration)"
+  }, limit = "3")
+  boolean equalsDuration(Object selfDuration, Object otherDuration,
+      @CachedLibrary("selfDuration") InteropLibrary selfInterop,
+      @CachedLibrary("otherDuration") InteropLibrary otherInterop) {
+    Duration selfJavaDuration;
+    Duration otherJavaDuration;
+    try {
+      selfJavaDuration = selfInterop.asDuration(selfDuration);
+      otherJavaDuration = otherInterop.asDuration(otherDuration);
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+    return selfJavaDuration.equals(otherJavaDuration);
+  }
+
+  /**
+   * Compares interop strings according to the lexicographical order, handling Unicode
+   * normalization. See {@code Text_Utils.compare_to}.
```

Comment:
We duplicate a (very small) part of code from `std-base` here.

For now - I think that is OK. But ideally, as it has been pointed out several times, we may want to have a way to share logic between Enso builtins and stdlib polyglot calls. We cannot only rely on builtins, as for example in the Table library we want to also compare strings without calling back to an Enso builtin - so we need to keep the `Text_Utils.compare_to`. But keeping this code in two places risks us getting out of sync.

It's probably not a big issue, especially as we test the text equality quite thoroughly, but it may be good to watch for this and figure something out if we get more situations like that.

Reply:
Righkt. In the long run, we ideally want to share `std-bits` between engine and std lib. I created [pivotal issue](https://www.pivotaltracker.com/story/show/184074597) to keep track of it.